### PR TITLE
feat: multicurrency

### PR DESCRIPTION
**Pull request by Eknir**
This PR adds the possibility to interact with the tokenContract with multiple currencies.

- The owner should whitelist each currency that the tokenContract can accept
- The owner is allowed to remove any currency from the whitelist at any moment. This means that no new offers and agreements can be made in that the non-whitelisted token. Running agreements will continue to run until expired
- the currency of the agreement is stored double mapping billingPlansForToken, where the first key is the whitelisted token
- the token reference with address(0) stands for the native currency

Note: the newAgreement event was shortened by removing `agreementReference`. This was necessary due to reaching the stack depth. An agreementReference can be acquired by calling the function `getAgreementReference` with the variables in the parameters of the event.

### CODE REVIEW

**Comment by julianmrodri** [comment]
@Eknir how are you going to manage if it is RBTC? Use the `zero` address?

File: contracts/StorageManager.sol [hunk 1]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
+    // maps the tokenAddresses which can be used with this contract. By convention, address(0) is the native token.
+    mapping(address => bool) public isWhitelistedToken;
+
     event TotalCapacitySet(address indexed provider, uint128 capacity);
-    event BillingPlanSet(address indexed provider, uint64 period, uint64 price);
+    event BillingPlanSet(address indexed provider, address token, uint64 period, uint64 price);
     event MessageEmitted(address indexed provider, bytes32[] message);
 
     event NewAgreement(
-        bytes32 agreementReference,
```
Comment (Eknir) [inline contracts/StorageManager.sol:54]: remove the datareference, add back the agreementreference

File: contracts/StorageManager.sol [hunk 2]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
     event AgreementStopped(bytes32 indexed agreementReference);
 
+    constructor() Ownable() public { }
+
+    /**
+    @notice whitelist a token or remove the token from whitelist
+    @param token the token from whom you want to set the whitelisted
+    @param isWhiteListed whether you want to whitelist the token or put it from the whitelist.
+    */
+    function setWhitelistedTokens(address token, bool isWhiteListed) public onlyOwner {
```
Comment (Eknir) [inline contracts/StorageManager.sol:80]: Note: this also allows for disabling a token @julianmrodri

Comment (Eknir) [reply contracts/StorageManager.sol:80]: Special attention is needed to verify that this doesn't create any edge conditions @vojtechsimetka

File: contracts/StorageManager.sol [hunk 3]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
-        uint64[] memory billingPrices,
+        uint64[][] memory billingPeriods,
+        uint64[][] memory billingPrices,
+        address[] memory tokens,
         bytes32[] memory message
     ) public {
         Offer storage offer = offerRegistry[msg.sender];
         setTotalCapacity(capacity);
-        require(billingPeriods.length > 0, "StorageManager: Offer needs some billing plans");
-        require(billingPeriods.length == billingPrices.length, "StorageManager: Billing plans array length has to equal to billing prices");
```
Comment (Eknir) [inline contracts/StorageManager.sol:87]: This check is already implicit because if not equal: `error: array index out of bounds`

File: contracts/StorageManager.sol [hunk 4]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
-        uint64[] memory billingPeriods,
-        uint64[] memory billingPrices,
+        uint64[][] memory billingPeriods,
+        uint64[][] memory billingPrices,
+        address[] memory tokens,
         bytes32[] memory message
     ) public {
         Offer storage offer = offerRegistry[msg.sender];
         setTotalCapacity(capacity);
-        require(billingPeriods.length > 0, "StorageManager: Offer needs some billing plans");
```
Comment (Eknir) [inline contracts/StorageManager.sol:86]: To verify @Eknir , why is this removed?

File: contracts/StorageManager.sol (hunk 3 above)
Comment (Eknir) [reply contracts/StorageManager.sol:87]: Not always true. Only true if one is longer than the other. Not an error if the other is longer than the one.

Comment (Eknir) [reply contracts/StorageManager.sol:87]: Should make a comment about it if not checked here.

File: contracts/StorageManager.sol [hunk 5]
```diff
@@ -308,23 +404,27 @@ contract StorageManager {
...
+                require(success, "StorageManager: Transfer failed.");
+            } else {
+                assert(ERC20(tokenOfAgreementsToPayOut).transfer(provider, toTransfer));
+            }
         }
     }
 
-    function getAgreementReference(bytes32[] memory dataReference, address creator) public pure returns (bytes32) {
-        return keccak256(abi.encodePacked(creator, dataReference));
+    function getAgreementReference(bytes32[] memory dataReference, address creator, address token) public pure returns (bytes32) {
```
Comment (Eknir) [inline contracts/StorageManager.sol:426]: @Eknir : TODO: make a comment

Comment (Eknir) [reply contracts/StorageManager.sol:426]: ln: 198

File: contracts/StorageManager.sol [hunk 6]
```diff
@@ -155,31 +173,55 @@ contract StorageManager {
...
-        bytes32 agreementReference = getAgreementReference(dataReference, msg.sender);
-        Agreement storage agreement = offer.agreementRegistry[agreementReference];
-
+        require(isWhitelistedToken[token], "StorageManager: not possible to interact witht this token");
+        // the agreementReference consists of the hash of the dataReference, msg.sender and the tokenAdddress, to allow:
+        // - multiple people to register an agreement for the same file
+        // - one person to register multiple agreements for the same file, but with different tokens
+        // - link the token to the agreement, such that we do the accounting properly
+        bytes32 agreementReference = getAgreementReference(dataReference, msg.sender, token);
+       // Agreement storage agreement = offer.agreementRegistry[agreementReference];
```
Comment (Eknir) [inline contracts/StorageManager.sol:203]: Remove these lines

Comment (Eknir) [reply contracts/StorageManager.sol:203]: Have a look if you can solve this in a better way.

File: contracts/StorageManager.sol [hunk 7]
```diff
@@ -155,31 +173,55 @@ contract StorageManager {
     @param provider the provider from which is proposed to take a Offer.
```
Comment (Eknir) [inline contracts/StorageManager.sol:173]: dev comment needed: if ERC20 token is used, first allow the contract to pull funds.

File: contracts/StorageManager.sol [hunk 8]
```diff
@@ -192,16 +234,15 @@ contract StorageManager {
 
         // Allow to enforce payout funds and close of agreements that are already expired,
         // which should free needed capacity, if the capacity is becoming depleted.
-        if (agreementsReferencesToBePayedOut.length > 0) {
-            _payoutFunds(agreementsReferencesToBePayedOut, payable(provider));
+        if (dataReferencesOfAgreementToPayOut.length > 0) {
+            _payoutFunds(dataReferencesOfAgreementToPayOut, creatorsOfAgreementToPayOut, tokenOfAgreementsToPayOut, payable(provider));
```
Comment (Eknir) [inline contracts/StorageManager.sol:238]: Move this one up into the function, before we are doing any accounting/check on our own agreement.

Comment (Eknir) [inline contracts/StorageManager.sol:238]: Idea: because we are only allowing payout of one token. Perhaps we should have a public function which does this as well, so we can still force payout of *all* agreements which are expired.

File: contracts/StorageManager.sol [hunk 9]
```diff
@@ -17,13 +20,13 @@ contract StorageManager {
     Offer represents:
      - utilizedCapacity: how much is capacity is utilized in Offer.
      - totalCapacity: total amount of bytes offered.
-     - billingPlans: maps a billing period to a billing price. When a price is 0, the period is not offered.
+     - billingPlansForToken: maps a whitelisted token to billing period to a billing price. When a price is 0, the period is not offered. Address 0 stands for the native currency. By convention, the 0-address stands for the native currency
```
Comment (julianmrodri) [inline contracts/StorageManager.sol:23]: Comment seems quite redundant the last 2 sentences I would leave only one.

File: contracts/StorageManager.sol [hunk 10]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
+
     event TotalCapacitySet(address indexed provider, uint128 capacity);
-    event BillingPlanSet(address indexed provider, uint64 period, uint64 price);
+    event BillingPlanSet(address indexed provider, address token, uint64 period, uint64 price);
     event MessageEmitted(address indexed provider, bytes32[] message);
 
     event NewAgreement(
-        bytes32 agreementReference,
-        bytes32[] dataReference,
+        bytes32[] dataReference,
```
Comment (julianmrodri) [inline contracts/StorageManager.sol:60]: can you explain this field? is this because you reached maximum size allow of info to emit in an Event? so to Obtain the `token` used for an agreement we would need to make a call after getting the event, right?

File: contracts/StorageManager.sol (hunk 2 above)
Comment (julianmrodri) [reply contracts/StorageManager.sol:80]: > Note: this also allows for disabling a token @julianmrodri

I think its OK!

File: contracts/StorageManager.sol (hunk 3 above)
Comment (Eknir) [reply contracts/StorageManager.sol:87]: I personally belief that Solidity should only contain verification, such that it ensures that the state is always consistent. Any business logic or seemingly surprising behavior for the user should be caught by the libraries or clients that interact with the contract. This is a good example of an unneeded check in Solidity code, as removing it doesn't risk bringing the contract in an invalid state. Nevertheless, we should add very explicitly to the `@dev` description on top of the function that billingPeriods length should be equal to billingPrices length.

Invalid input should then be caught by the client.

File: contracts/StorageManager.sol (hunk 4 above)
Comment (Eknir) [reply contracts/StorageManager.sol:86]: Ha @julianmrodri , I am talking to myself as I was reviewing the changeset. It was removed but while removing couldn't catch why exactly. I'll review this and report here.

File: contracts/StorageManager.sol [hunk 11]
```diff
@@ -1,12 +1,15 @@
...
 import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
 
 /// @title StorageManager
 /// @author Rinke Hendriksen <rinke@iovlabs.org>
 /// @author Adam Uhlir <adam@iovlabs.org>
 /// @notice Providers can offer their storage space and list their price and Consumers can take these offers
-contract StorageManager {
+contract StorageManager is Ownable {
```
[the remark this answers is not in the document]
Comment (Eknir) [reply contracts/StorageManager.sol:12]: I don't remember this, it was not in my TODO list at least. Can be added of course. Could you have a look and decide which functions you want to Pause?

File: contracts/StorageManager.sol [hunk 12]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
     );
-    event AgreementFundsDeposited(bytes32 indexed agreementReference, uint256 amount);
-    event AgreementFundsWithdrawn(bytes32 indexed agreementReference, uint256 amount);
-    event AgreementFundsPayout(bytes32 indexed agreementReference, uint256 amount);
+    event AgreementFundsDeposited(bytes32 indexed agreementReference, uint256 amount, address indexed token);
+    event AgreementFundsWithdrawn(bytes32 indexed agreementReference, uint256 amount, address indexed token);
+    event AgreementFundsPayout(bytes32 indexed agreementReference, uint256 amount, address indexed token);
     event AgreementStopped(bytes32 indexed agreementReference);
 
+    constructor() Ownable() public { }
```
[the remark this answers is not in the document]
Comment (Eknir) [reply contracts/StorageManager.sol:73]: We will consider upgradability separately. Thanks for the info though ;)

File: contracts/StorageManager.sol [hunk 13]
```diff
@@ -1,12 +1,15 @@
 pragma solidity 0.6.2;
+pragma experimental ABIEncoderV2;
 
 import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
```
[the remark this answers is not in the document]
Comment (Eknir) [reply contracts/StorageManager.sol:6]: Why?

File: contracts/StorageManager.sol [hunk 14]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
+    }
+
+
     /**
     >> FOR PROVIDER
     @notice set the totalCapacity and billingPlans of a Offer.
     @dev
     - Use this function when initiating an Offer or when the users wants to change more than one parameter at once.
     - make sure that any period * prices does not cause an overflow, as this can never be accepted (REF_MAX_PRICE) and hence is pointless
+    - only whitelisted tokens are allowed to make a token for
```
[the remark this answers is not in the document]
Comment (Eknir) [reply contracts/StorageManager.sol:90]: You are correct, thanks for spotting. I'll update

File: contracts/StorageManager.sol (hunk 9 above)
Comment (Eknir) [reply contracts/StorageManager.sol:23]: Correct. Will remove. Thx

File: contracts/StorageManager.sol (hunk 10 above)
Comment (Eknir) [reply contracts/StorageManager.sol:60]: See: https://github.com/rsksmart/rif-marketplace-storage/pull/113#discussion_r471940742
Probably made a mistake here. Pls have a look after I commit changes if you agree

File: contracts/StorageManager.sol (hunk 4 above)
Comment (Eknir) [reply contracts/StorageManager.sol:86]: I verified. If the billingPeriod for a particular token is not mentioned, the token will just be disregarded and no state changes will be made (line 372).

File: contracts/StorageManager.sol (hunk 6 above)
Comment (Eknir) [reply contracts/StorageManager.sol:202]: Removed this. Cannot do it in a better way. Compiler error: stack too deep.

File: contracts/StorageManager.sol [hunk 15]
```diff
@@ -244,54 +290,103 @@ contract StorageManager {
...
+            }
+            require(amount <= maxWithdrawableFunds, "StorageManager: Amount is too big");
+            agreement.availableFunds = agreement.availableFunds.sub(amount);
+            require(amount > 0, "StorageManager: Nothing to withdraw");
+
+            if(token == address(0)) {
+                (bool success,) = msg.sender.call.value(amount)("");
+                require(success, "Transfer failed.");
+            } else {
+                require(ERC20(token).transferFrom(msg.sender, address(this), amount), "StorageManager: not allowed to deposit tokens from token contract");
```
Comment (Eknir) [inline contracts/StorageManager.sol:325]: Needs to be transfer

File: contracts/StorageManager.sol (hunk 3 above)
Comment (Eknir) [reply contracts/StorageManager.sol:87]: TODO @Eknir , add this comment

File: contracts/StorageManager.sol (hunk 1 above)
Comment (Eknir) [reply contracts/StorageManager.sol:54]: Done

File: contracts/StorageManager.sol (hunk 4 above)
Comment (Eknir) [reply contracts/StorageManager.sol:86]: Done

File: contracts/StorageManager.sol (hunk 5 above)
Comment (Eknir) [reply contracts/StorageManager.sol:435]: Done

File: contracts/StorageManager.sol (hunk 7 above)
Comment (Eknir) [reply contracts/StorageManager.sol:173]: Done

File: contracts/StorageManager.sol (hunk 8 above)
Comment (Eknir) [reply contracts/StorageManager.sol]: Done

File: contracts/StorageManager.sol (hunk 14 above)
[the remark this answers is not in the document]
Comment (Eknir) [reply contracts/StorageManager.sol]: Done

File: contracts/StorageManager.sol (hunk 9 above)
Comment (Eknir) [reply contracts/StorageManager.sol]: Done

File: contracts/StorageManager.sol (hunk 15 above)
Comment (Eknir) [reply contracts/StorageManager.sol]: Done.

File: contracts/StorageManager.sol (hunk 8 above)
Comment (Eknir) [reply contracts/StorageManager.sol]: Done: This was surprisingly easy to do. We add param provider to the public function payoutFunds.
@julianmrodri , this is a new feature, essentially to allow people to free capacity from expired agreements. We do this also in the function `newAgreement`. However, compared to the previous version, newAgreement only allows to do a payout for a single token. If you would like to force payout of expired agreements otherwise, now this is possible via this public function.

File: contracts/StorageManager.sol [hunk 16]
```diff
@@ -46,48 +49,63 @@ contract StorageManager {
...
+    }
+
     /**
     >> FOR PROVIDER
     @notice set the totalCapacity and billingPlans of a Offer.
     @dev
     - Use this function when initiating an Offer or when the users wants to change more than one parameter at once.
     - make sure that any period * prices does not cause an overflow, as this can never be accepted (REF_MAX_PRICE) and hence is pointless
+    - only whitelisted tokens are allowed to make an offer for
+    - if there are two tokens, and two billingPrice/periods pairs per token, then boundaries[0] == 1.
```
Comment (Eknir) [inline contracts/StorageManager.sol:91]: TODO @Eknir . This is outdated

File: contracts/StorageManager.sol [hunk 17]
```diff
@@ -122,18 +140,20 @@ contract StorageManager {
     @notice set the billing plans for an Offer.
     @dev
     - setting the price to 0 means that a particular period is not offered, which can be used to remove a period from the offer.
-    - make sure that any period * prices does not cause an overflow, as this can never be accepted (REF_MAX_PRICE) and hence is pointless.
-    @param billingPeriods the offered periods. Length must be equal to billingPrices.
-    @param billingPrices the prices for the offered periods. Each entry at index corresponds to the same index at periods. 0 means that the particular period is not offered.
+    - make sure that any period * prices does not cause an overflow, as this can never be accepted and hence is pointless.
+    - the length of tokens array must always be one shorter than the length of the boundaries array (otherwise you get an array index out of bounds error)
```
Comment (Eknir) [inline contracts/StorageManager.sol:144]: TODO @Eknir , this is outdated

File: contracts/StorageManager.sol [hunk 18]
```diff
@@ -242,56 +290,108 @@ contract StorageManager {
...
+    ) public {
+        _payoutFunds(dataReferencesOfAgreementToPayOut, creatorsOfAgreementToPayOut, tokensOfAgreementsToPayOut, provider);
     }
 
-    function _payoutFunds(bytes32[] memory agreementReferences, address payable provider) internal {
+    /**
+    @notice sets the billing plans for multiple tokens. 
+    @dev
+    - the billingPeriods and billingPrices hold the period/price pair for all tokens. 
+    - the length of tokens array must always be one shorter than the length of the boundaries array (otherwise you get an array index out of bounds error)
```
Comment (Eknir) [inline contracts/StorageManager.sol:360]: TODO @Eknir , this is outdated

File: contracts/StorageManager.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/StorageManager.sol:6]: AFAIK when deploying through Proxy you need the versions that have the `initialize` function instead of constructors.  The package they recommend to be used for Upgradable contracts is: https://github.com/OpenZeppelin/openzeppelin-contracts-ethereum-package. "  
There is an explanation here: https://docs.openzeppelin.com/upgrades/2.8/writing-upgradeable and they have an example with ERC20, but I guess in our case we are not instantiating or deploying those so maybe its not required. This seems required if you plan to get the constructors of the contracts executed or if you expect to upgrade those contracts.

File: contracts/StorageManager.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/StorageManager.sol:12]: I think if we Pause the contract we probably want to disable the creation of any Offer, Agreement and Deposit of new funds (Like renewing agreements) All other should continue (allow to withdraw for example). what do you think?

File: contracts/StorageManager.sol [hunk 19]
```diff
@@ -1,12 +1,15 @@
 pragma solidity 0.6.2;
+pragma experimental ABIEncoderV2;
 
 import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
```
Comment (julianmrodri) [inline contracts/StorageManager.sol:5]: Actually I think here would be enough to just load the interface `IERC20.sol` right?
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/IERC20.sol

File: contracts/StorageManager.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (Eknir) [reply contracts/StorageManager.sol:6]: Ah yes. No, in this case it is indeed not needed, as we are only using the ERC20 for the interface (so our contract knows how to interact with it). Besides, this PR does not add upgradability.

File: contracts/StorageManager.sol (hunk 11 above)
[the remark this answers is not in the document]
Comment (julianmrodri) [reply contracts/StorageManager.sol:12]: @nduchak

File: contracts/StorageManager.sol (hunk 19 above)
Comment (julianmrodri) [reply contracts/StorageManager.sol:5]: @nduchak

Comment (nduchak) [reply contracts/StorageManager.sol]: Done

File: contracts/StorageManager.sol [hunk 20]
```diff
@@ -46,48 +50,63 @@ contract StorageManager {
...
     );
-    event AgreementFundsDeposited(bytes32 indexed agreementReference, uint256 amount);
-    event AgreementFundsWithdrawn(bytes32 indexed agreementReference, uint256 amount);
-    event AgreementFundsPayout(bytes32 indexed agreementReference, uint256 amount);
+    event AgreementFundsDeposited(bytes32 indexed agreementReference, uint256 amount, address indexed token);
+    event AgreementFundsWithdrawn(bytes32 indexed agreementReference, uint256 amount, address indexed token);
+    event AgreementFundsPayout(bytes32 indexed agreementReference, uint256 amount, address indexed token);
     event AgreementStopped(bytes32 indexed agreementReference);
 
+    constructor() Ownable() Pausable() public { }
```
Comment (julianmrodri) [inline contracts/StorageManager.sol:74]: Never seen this usage, is it setting correctly the deployer address as owner and pauser?

Comment (julianmrodri) [reply contracts/StorageManager.sol:74]: Anyway all this will change with Upgradability as constructor will be removed with that change. So doesnt matter too much.

File: contracts/StorageManager.sol [hunk 21]
```diff
@@ -242,56 +292,113 @@ contract StorageManager {
...
     */
-    function payoutFunds(bytes32[] memory agreementReferences) public {
-        _payoutFunds(agreementReferences, msg.sender);
+
+    function payoutFunds(
+        bytes32[][] memory dataReferencesOfAgreementToPayOut,
+        address[] memory creatorsOfAgreementToPayOut,
+        address tokensOfAgreementsToPayOut,
+        address payable provider
+    ) public whenNotPaused {
```
Comment (julianmrodri) [inline contracts/StorageManager.sol:359]: we should allow Withdrawals when Paused.

File: contracts/StorageManager.sol [hunk 22]
```diff
@@ -242,56 +292,113 @@ contract StorageManager {
...
+    @param tokens the tokens in which to withdraw. By convention, address(0) is the native currency.
+    @param amounts the value you want to withdraw for each token
     */
-    function withdrawFunds(bytes32[] memory dataReference, address provider, uint256 amount) public payable {
+    function withdrawFunds(
+        bytes32[] memory dataReference,
+        address provider,
+        address[] memory tokens,
+        uint256[] memory amounts
+    ) public whenNotPaused {
```
Comment (julianmrodri) [inline contracts/StorageManager.sol:306]: we should allow to withdraw even if its paused.

**Review by julianmrodri — CHANGES_REQUESTED** [review]
LG!! Just sent some comments about allowing to withdraw when paused, and adding test for Pausability.

File: contracts/StorageManager.sol (hunk 22 above)
Comment (nduchak) [reply contracts/StorageManager.sol]: Done

File: contracts/StorageManager.sol (hunk 21 above)
Comment (nduchak) [reply contracts/StorageManager.sol]: Done

File: test/StorageManager.test.js [hunk 23]
```diff
@@ -621,4 +621,76 @@ contract('StorageManager', ([Provider, Consumer, randomPerson]) => {
...
+      expect(await storageManager.paused()).to.be.eql(true)
+      const msg = [padRight(asciiToHex('some string'), 64), padRight(asciiToHex('some other string'), 64)]
+      await expectRevert(
+        storageManager.setOffer(1000, [[10, 100], [20, 100]], [[10, 80], [20, 80]], [constants.ZERO_ADDRESS, token.address], msg, { from: Provider }),
+        'Pausable: paused'
+      )
+    })
+    it('should not be able to to set capacity when paused', async () => {
+      await storageManager.setOffer(1000, [[10, 100], [20, 100]], [[10, 80], [20, 80]], [constants.ZERO_ADDRESS, token.address], [], { from: Provider })
+      await storageManager.pause({ from: randomPerson })
```
Comment (julianmrodri) [inline test/StorageManager.test.js:637]: would rename the varilable as `randomPerson` looks like somebody that should NOT be allowed to pause. Maybe we can create an `owner` variable and use that for the owner in the tests to make it more clear when reading.

File: contracts/StorageManager.sol [hunk 24]
```diff
@@ -129,7 +129,7 @@ contract StorageManager is Ownable, Pausable {
     @notice stops the Offer. It sets the totalCapacity to 0 which indicates terminated Offer.
     @dev no new Agreement can be created and no existing Agreement can be prolonged. All existing Agreement are still valid for the amount of periods still deposited.
     */
-    function terminateOffer() public {
+    function terminateOffer() public whenNotPaused {
```
Comment (julianmrodri) [inline contracts/StorageManager.sol:132]: I would not put it here. I think we should allow providers to `cleanup` and make sure they can disable future agreements when contract is resumed. Would give them more safety, on the same lines as allowing all parties to withdraw funds even if paused.

**Review by julianmrodri — CHANGES_REQUESTED** [review]
LG! only two items to review.

File: contracts/StorageManager.sol (hunk 24 above)
Comment (nduchak) [reply contracts/StorageManager.sol]: Done

File: test/StorageManager.test.js (hunk 23 above)
Comment (nduchak) [reply test/StorageManager.test.js]: Done

**Review by julianmrodri — APPROVED** [review]
LGTM
